db-engine/src/main/java/com/splicemachine/db/impl/jdbc/authentication/TokenAuthenticationServiceImpl.java

```diff
@@ -0,0 +1,136 @@
+ * </p>
+ *
+ * <p>
+ * To activate this service, set the db.authentication.provider database
+ * or system property to a value beginning with the token "TOKEN:".
+ * </p>
+ *
+ * <p>
+ * This service instantiates and calls the basic User authentication scheme at runtime.
+ * </p>
+ *
+ * <p>
+ * User credentials are defined in the SYSUSERS table.
+ * </p>
+ */
+public final class TokenAuthenticationServiceImpl extends NativeAuthenticationServiceImpl {
+
+    public String authenticateUser (String userName, String userPassword, String databaseName, Properties info) throws SQLException {
+        if (userPassword != null) {
+            return super.authenticateUser(userName, userPassword, databaseName, info);
+        } else {
+            return authenticateUserWithToken(userName, info);
+        }
+    }
+
+    private String authenticateUserWithToken (String userName, Properties info) throws SQLException {
+        try {
+            String jwt = info.getProperty(Attribute.USER_TOKEN);
+            if (jwt == null) {
+                return null;
+            }
+
+            String tokenUserName = getTokenVerifier(info.getProperty(Attribute.USER_TOKEN_AUTHENTICATOR)).decodeUsername(jwt);
+
+            if (!tokenUserName.equals(IdUtil.getUserAuthorizationId(userName))) {
+                return null;
+            }
+
+            if (userName == null) {
+                return null;
+            }
+
+            if (!authenticateLocally(userName)) {
+                return null;
+            }
+
+            // check group mapping
+            List<String> groupList = new ArrayList<>();
+            groupList.add(userName);
+            return mapUserGroups(groupList);
+        } catch (StandardException se) {
+            throw Util.generateCsSQLException(se);
+        }
+    }
+
+
+    /**
+     * Authenticate the passed-in credentials against the local database.
+     *
+     * @param userName     The user's name used to connect to JBMS system
+     */
+    private boolean authenticateLocally(String userName) throws StandardException {
+        userName = IdUtil.getUserAuthorizationId(userName);
+
+        //
+        // we expect to find a data dictionary
+        //
+        DataDictionary dd = (DataDictionary) Monitor.getServiceModule(this, DataDictionary.MODULE);
+
+        UserDescriptor userDescriptor = dd.getUser(userName);
+
+        return userDescriptor != null;
+    }
+
+    private AccessTokenVerifier getTokenVerifier(String authenticatorName) throws StandardException {
+        try {
+            Class<?> verifierFactoryClass = Class.forName("com.splicemachine.db.impl.token.AccessTokenVerifierFactory");
+            Method m = verifierFactoryClass.getDeclaredMethod("createVerifier", String.class);
+            return (AccessTokenVerifier) m.invoke(null, authenticatorName);
+        } catch (Exception e) {
+            throw StandardException.plainWrapException(e);
+        }
```

Comment:
I don't like this reflection here. Would it be possible to use something similar to the BackupManager? A class in the -ee repo implements the AuthManager, a class in this repo throws UnsupportedOpEx.

Reply:
Or even move TokenAuthenticationServiceImpl to the -ee repo?

Reply:
TokenAuthenticationServiceImpl has been moved to EE